Update/variations search method

joshuatf opened the pull request:
Fixes #1773 

* Adds `product_search` to variations controller.
* Hooks up `product_search` to variation autocompleters.
* Adds product sku to autocompleter keywords.

### Screenshots
<img width="633" alt="Screen Shot 2019-03-13 at 5 22 57 PM" src="https://user-images.githubusercontent.com/10561050/54267777-c864f480-45b4-11e9-9187-d5cc692b96ea.png">

### Detailed test instructions:

1. Create a variable product with many products and varying skus.
2. Check that the autocompleter options correctly return products by searched name/sku.

@psealock Can you double check and let me know if this works with your data?

joshuatf committed:
Add product search option REST API

joshuatf committed:
Add product search option REST API

joshuatf committed:
Remove product_name search in favor of sku and name searching

joshuatf committed:
Remove product_name search in favor of sku and name searching

joshuatf committed:
Search by product search term instead of generic search in autocompleter

joshuatf committed:
Add product search param to variations controller

joshuatf committed:
Add group by to prevent duplicate variations in response

joshuatf committed:
Use product_search in variation autocompleter

psealock reviewed (commented):
Indeed I am able to search variation names in the table as well as by SKU. Nice work.

1. We can change the input placeholder back to `Search by variation name or SKU`.
2. I ran into one unrelated issue when searching for a Single Product:

![Screen Shot 2019-03-14 at 2 07 38 PM](https://user-images.githubusercontent.com/1922453/54324576-1f290780-4663-11e9-8da0-26082ca82c0e.png)

It occurs on master as well, I'm just noting it here in case you are also experiencing the same thing.

3. One related issue is when selecting the search option, the url removes the `?filter=single_product` parameter. I ran into this problem when looking at categories and [solved it](https://github.com/woocommerce/woocommerce-admin/pull/1762) by introducing a table prop called `baseSearchQuery`

https://github.com/woocommerce/woocommerce-admin/blob/80a00415497f4f28340d0331608460a3ddd93fa6/client/analytics/report/categories/index.js#L84

joshuatf committed:
Trim whitespace in product_search query

joshuatf committed:
Change variation search placeholder

joshuatf commented:
@psealock Thanks for the review!

1. Updated.
2. Did not run into this issue here or master.  Any filters in place that cause this?
3. I might be misunderstanding the issue, but if we're using a single product filter and subsequently use the search, I would expect the filter to be removed since you wouldn't want to search a single product.

jeffstieler commented on includes/api/class-wc-admin-rest-product-variations-controller.php:
```diff
@@ -22,4 +22,69 @@ class WC_Admin_REST_Product_Variations_Controller extends WC_REST_Product_Variat
+	 */
+	public function get_collection_params() {
+		$params                   = parent::get_collection_params();
+		$params['product_search'] = array(
```
Same question as https://github.com/woocommerce/woocommerce-admin/pull/1791/files#r265597768:

Do we have to use `product_search` instead of `search` here? Since we're already overriding `get_collection_params()`, can't we handle the `search` parameter there?

joshuatf replied:
@jeffstieler I was wondering the same.  My concern was that other plugins taking advantage of the `search` param might run into unexpected behavior, but maybe this is okay given that we've namespaced the API to v4.

jeffstieler replied:
Yeah I think the namespacing saves us here - we're changing things much more drastically in the customers endpoint: https://github.com/woocommerce/woocommerce-admin/blob/master/includes/api/class-wc-admin-rest-customers-controller.php#L18

psealock commented:
>if we're using a single product filter and subsequently use the search, I would expect the filter to be removed since you wouldn't want to search a single product.

Thats true, except in this case we are searching for variations of a particular product (I removed the chart in all these so it fits nicely):

Notice the title of the table is "Variations" and we are looking at the single product Penguin.

![Screen Shot 2019-03-15 at 9 26 35 AM](https://user-images.githubusercontent.com/1922453/54389256-8a76e600-4704-11e9-9a69-b89af6a6d4ac.png)

When I click a search selection, because the top level filter `filter=single_product` is removed from the url, we go back to looking all products, and no longer variations of that single product. I see what you are saying, but the context of the user's search was in regards to the single product, so I think the resulting view should reflect that.

![Screen Shot 2019-03-15 at 9 29 09 AM](https://user-images.githubusercontent.com/1922453/54389431-ec375000-4704-11e9-90d0-e847cbcc23ee.png)

When I manually add back in the `filter=single_product` param, the filter is correct, but the table is not right. It should be the variations table and not the product.

![Screen Shot 2019-03-15 at 9 32 31 AM](https://user-images.githubusercontent.com/1922453/54389602-5819b880-4705-11e9-8987-0752799b9abf.png)

psealock commented:
>I ran into one unrelated issue when searching for a Single Product:

>>Did not run into this issue here or master. Any filters in place that cause this?

Weird, I will dig into why I'm seeing this.

joshuatf committed:
Override existing search param instead of new param

joshuatf committed:
Override search param instead of using new product_search param

joshuatf committed:
Add baseSearchQuery to prevent losing single_product filter in variations report

joshuatf commented:
>Thats true, except in this case we are searching for variations of a particular product.

@psealock You're absolutely right; I was thinking about this the wrong way.  Added `baseSearchQuery` in https://github.com/woocommerce/woocommerce-admin/pull/1792/commits/78fe98772b781b51d9afcd5b5f4a084a431f00ff

@jeffstieler Rebased off `update/wp-query-product-search` and changed params to use `search` instead of `product_search`.

jeffstieler commented on includes/api/class-wc-admin-rest-product-variations-controller.php:
```diff
@@ -22,4 +22,70 @@ class WC_Admin_REST_Product_Variations_Controller extends WC_REST_Product_Variat
+
+		if ( ! empty( $request['search'] ) ) {
+			$args['search'] = $request['search'];
+			$args['s']      = false;
```
Nitpick: I think it would be better to `unset()` here. _(`false` still passes an `isset()` check that other code might be using)_

jeffstieler commented on includes/api/class-wc-admin-rest-product-variations-controller.php:
```diff
@@ -22,4 +22,70 @@ class WC_Admin_REST_Product_Variations_Controller extends WC_REST_Product_Variat
+
+		$search = trim( $wp_query->get( 'search' ) );
+		if ( $search && empty( $groupby ) ) {
+			$groupby .= $wpdb->posts . '.ID';
```
Nitpick: since we know `$groupby` is empty, there's no need for concatenation.

jeffstieler reviewed (approved):
Code looks good and tests well - optionally handle two nitpicks before merging.

joshuatf committed:
Unset s param instead of setting to false
